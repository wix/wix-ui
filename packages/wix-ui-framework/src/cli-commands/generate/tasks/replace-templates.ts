import * as ejs from 'ejs';

import * as camelCase from 'lodash/camelCase';
import * as kebabCase from 'lodash/kebabCase';
import * as snakeCase from 'lodash/snakeCase';

const utils = {
  toCamel: camelCase,
  toKebab: kebabCase,
  toSnake: snakeCase,
  toPascal: (s: string) => {
    const camel: string = camelCase(s);
    return camel[0].toUpperCase() + camel.substring(1);
  },
};

export const replaceTemplates = (source: string, scope: Object) => {
  /*
   * TODO: usage of `{%` & `%}` tags is deprecated!
   * regular EJS should be used in user templates
   *
   * keeping it to not break existing users
   * can be removed either:
   * 1. with major version release
   * 2. when ensured no consumers use it
   */
  const replaceDeprecatedSyntax = source.replace(/{%[\w-]+%}/g, (match) => {
    const key = match.slice(2, -2);

    if (!scope.hasOwnProperty(key)) {
      throw new Error(`key '${key}' must be one of: [${Object.keys(scope)}]`);
    }

    return scope[key];
  });

  try {
    return ejs.render(replaceDeprecatedSyntax, {
      utils,
      ...scope,
    });
  } catch (e) {
    throw new Error(`Erroneous template syntax: ${e}`);
  }
};
