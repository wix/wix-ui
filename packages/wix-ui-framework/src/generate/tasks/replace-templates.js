// Extracted from
// https://github.com/wix/yoshi/blob/master/packages/create-yoshi-app/src/replaceTemplates.js

const templateRegex = /{%[\w-]+%}/g;

module.exports = (content, map) => {
  function replacer(match) {
    const key = match.slice(2, -2);

    if (!map.hasOwnProperty(key)) {
      throw new Error(`key '${key}' must be one of: [${Object.keys(map)}]`);
    }

    return map[key];
  }

  return content.replace(templateRegex, replacer);
};
