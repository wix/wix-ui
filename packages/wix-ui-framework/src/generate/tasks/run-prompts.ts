import * as prompts from 'prompts';

import { isPascalCase } from '../utils';
const logger = require('../logger');

export const runPrompts = async () => {
  let promptAborted = false;

  const questions = [
    {
      type: 'text',
      name: 'ComponentName',
      message: 'Component name (PascalCase)',
      validate: value => {
        if (!value.length) {
          return 'Please supply a component name';
        }

        if (!isPascalCase(value)) {
          return 'Component name must be in PascalCase';
        }

        return true;
      },
    },
    {
      type: 'text',
      name: 'description',
      message: 'Description',
    },
  ];

  const answers = await prompts(questions, {
    onCancel: () => {
      promptAborted = true;
    },
  });

  if (promptAborted) {
    logger.divider();
    logger.error('Aborted.');
    process.exit(1);
  }

  return answers;
};
