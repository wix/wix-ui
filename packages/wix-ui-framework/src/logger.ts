import chalk from 'chalk';
import ora from 'ora';

export const error = (msg: string) => console.log(`${chalk.red('✖')} ${msg}`);
export const success = (msg: string) =>
  console.log(`${chalk.green('✔')} ${msg}`);
export const info = (msg: string) => console.log(`${chalk.blue('ℹ')} ${msg}`);
export const divider = () => console.log();
export const spinner = (msg: string) => ora(msg).start();
