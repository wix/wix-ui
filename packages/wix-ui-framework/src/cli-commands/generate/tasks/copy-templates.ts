import * as fs from 'fs';
import * as path from 'path';
import * as mkdirp from 'mkdirp';

import { Options } from '../typings';
import { replaceTemplates } from './replace-templates';

import { createValuesMap } from '../create-values-map';

const templateNamePlaceholder = 'Component';

const pathExists = p =>
  new Promise(resolve => {
    fs.access(p, fs.constants.F_OK, err => {
      resolve(!err);
    });
  });

export const copyTemplates: (a: Options) => Promise<void> = async ({
  templates,
  ComponentName,
  description,
  _process,
  output = '',
}) => {
  if (!ComponentName) {
    throw new Error('Component name must be provided!');
  }

  const readdir = (entry: string) =>
    fs.readdirSync(entry).map(dir => path.join(entry, dir));

  const queue = readdir(templates);

  while (queue.length) {
    const itemPath = queue.shift();

    const isDir = fs.lstatSync(itemPath).isDirectory();

    const newDestPath = path
      .join(_process.cwd, output, itemPath.replace(templates, ''))
      .replace(new RegExp(`${templateNamePlaceholder}`, 'g'), ComponentName);

    if (isDir) {
      queue.push(...readdir(path.join(itemPath)));

      if (!(await pathExists(newDestPath))) {
        mkdirp.sync(newDestPath);
      }
    } else if (!(await pathExists(newDestPath))) {
      const replaced = replaceTemplates(
        fs.readFileSync(itemPath, 'utf8'),
        createValuesMap({ ComponentName, description }),
      );

      fs.writeFileSync(newDestPath, replaced, 'utf8');
    }
  }
};
