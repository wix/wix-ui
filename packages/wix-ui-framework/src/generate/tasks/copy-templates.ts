import * as fs from 'fs';
import * as path from 'path';

import { Options } from '../typings';
import { replaceTemplates } from './replace-templates';

import { createValuesMap } from '../create-values-map';

const templateNamePlaceholder = 'Component';

const pathExists = p =>
  new Promise(resolve => {
    fs.access(p, fs.constants.F_OK, err => {
      resolve(!err);
    });
  });

export const copyTemplates: (a: Options) => void = async ({
  cwd,
  templates,
  ComponentName,
  description,
}) => {
  if (!ComponentName) {
    throw new Error('Component name must be provided!');
  }

  const readdir = entry =>
    fs.readdirSync(entry).map(dir => path.join(entry, dir));

  const queue = readdir(templates);

  while (queue.length) {
    const itemPath = queue.shift();

    const isDir = fs.lstatSync(itemPath).isDirectory();

    const newDestPath = path
      .join(cwd, itemPath.replace(templates, ''))
      .replace(new RegExp(`${templateNamePlaceholder}`, 'g'), ComponentName);

    if (isDir) {
      queue.push(...readdir(path.join(itemPath)));

      if (!(await pathExists(newDestPath))) {
        fs.mkdirSync(newDestPath);
      }
    } else if (!(await pathExists(newDestPath))) {
      const replaced = replaceTemplates(
        fs.readFileSync(itemPath, 'utf8'),
        createValuesMap({ ComponentName, description }),
      );

      fs.writeFileSync(newDestPath, replaced, 'utf8');
    }
  }
};
