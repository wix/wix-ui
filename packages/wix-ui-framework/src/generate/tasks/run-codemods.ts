import * as fs from 'fs';
import * as path from 'path';
import { exec } from 'child_process';

import * as logger from '../logger';
import { Options, CodemodConfig } from '../typings.d';
import { createValuesMap, CodemodValues } from '../create-values-map';

const runCodemod: ({
  options: Options,
  codemodValues: CodemodValues,
  codemodConfig: CodemodConfig,
}) => Promise<void> = ({ options, codemodValues, codemodConfig }) =>
  new Promise((resolve, reject) => {
    const codemodPath = path.join(options.codemods, codemodConfig.codemod);

    const pathToExecutable = path.join(
      options.cwd,
      'node_modules',
      '.bin',
      'jscodeshift',
    );

    const inputPath = path.join(options.cwd, codemodConfig.dist);

    if (!fs.existsSync(inputPath)) {
      reject(`Error in ${codemodConfig.codemod}: ${inputPath} does not exist`);
      return;
    }

    const command = `${pathToExecutable} \
      ${inputPath} \
      -t ${codemodPath} \
      --ComponentName=${codemodValues.ComponentName} \
      --componentName=${codemodValues.componentName} \
      --verbose=2`;

    const execProc = exec(command);

    execProc.stderr.on('data', data => {
      logger.error(`Error in ${codemodConfig.codemod}: ${data.toString()}`);
      reject(data.toString());
    });

    execProc.on('exit', () => {
      logger.success(codemodConfig.description);
      resolve();
    });
  });

export const runCodemods: (a: Options) => Promise<void[]> = options => {
  const { ComponentName, description, codemods } = options;

  const codemodsIndex: CodemodConfig[] = require(path.join(
    codemods,
    'index.js',
  ));

  const codemodValues: CodemodValues = createValuesMap({
    ComponentName,
    description,
  });

  return Promise.all(
    codemodsIndex.map((codemodConfig: CodemodConfig) =>
      runCodemod({
        options,
        codemodValues,
        codemodConfig,
      }),
    ),
  );
};
