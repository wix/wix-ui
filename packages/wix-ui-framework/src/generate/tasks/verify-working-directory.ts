import * as logger from '../logger';
import * as utils from '../utils';

export const errorMessage =
  'Git working directory is dirty!\nCommit or stash your changes, or run the generator with the --force flag';

export const verifyWorkingDirectory = async ({ cwd }) => {
  if (!(await utils.isGitRepoClean(cwd))) {
    logger.divider();
    logger.divider();
    logger.error(errorMessage);
    logger.divider();
    process.exit(1);
  }
};
