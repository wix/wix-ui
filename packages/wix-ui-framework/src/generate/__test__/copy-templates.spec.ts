import * as fs from 'fs';
import * as path from 'path';
import * as tempy from 'tempy';
import globby from 'globby';
import * as logger from '../logger';
import { copyTemplates } from '../tasks/copy-templates';

// Extracted from
// https://github.com/wix/yoshi/blob/master/packages/create-yoshi-app/src/getFilesInDir.js
const getDirSnapshot = absoluteDirPath => {
  const filesPaths = globby.sync(['**/*', '!node_modules'], {
    cwd: absoluteDirPath,
    dot: true,
    gitignore: true,
  });

  const files = {};

  filesPaths.forEach(filePath => {
    const content = fs.readFileSync(
      path.join(absoluteDirPath, filePath),
      'utf-8',
    );

    files[filePath] = content;
  });

  return files;
};

describe('copyTemplates', () => {
  let succesSpy;
  let tempDir;

  beforeEach(() => {
    tempDir = tempy.directory();

    // Silent logs
    succesSpy = jest.spyOn(logger, 'success').mockImplementation(() => {});
  });

  afterEach(() => {
    succesSpy.mockRestore();
  });

  it('should work as expected when description is provided', async () => {
    await copyTemplates({
      ComponentName: 'MyNewComponent',
      description: "This is a very cool component, ya'll",
      templatesPath: path.join(__dirname, 'templates'),
      cwd: tempDir,
    });

    expect(getDirSnapshot(tempDir)).toMatchSnapshot();
  });

  it('should work as expected when description is not provided', async () => {
    await copyTemplates({
      ComponentName: 'MyNewComponent',
      description: undefined,
      templatesPath: path.join(__dirname, 'templates'),
      cwd: tempDir,
    });

    expect(getDirSnapshot(tempDir)).toMatchSnapshot();
  });
});
