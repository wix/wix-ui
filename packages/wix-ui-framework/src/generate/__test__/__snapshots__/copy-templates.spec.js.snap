// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`copyTemplates should work as expected when description is not provided 1`] = `
Object {
  "test-generated/src/MyNewComponent/MyNewComponent.js": "import React from 'react';
import PropTypes from 'prop-types';

import Text from '../Text';
import Button from '../Button';
import styles from './MyNewComponent.st.css';


class MyNewComponent extends React.PureComponent {
  static displayName = 'MyNewComponent';

  static propTypes = {
    dataHook: PropTypes.string,

    /** Text for the button */
    buttonText: PropTypes.string,
  };

  static defaultProps = {
    buttonText: 'Click me!',
  };

  state = {
    count: 0,
  };

  _handleClick = () => {
    this.setState(({ count }) => ({
      count: count + 1,
    }));
  };

  render() {
    const { count } = this.state;
    const { dataHook, buttonText } = this.props;
    const isEven = count % 2 === 0;

    return (
      <div className={styles.root} data-hook={dataHook}>
        <Text dataHook=\\"myNewComponent-count\\">
          You clicked this button {isEven ? 'even' : 'odd'} number (
          <span
            {...styles('number', { even: isEven, odd: !isEven }, this.props)}
          >
            {count}
          </span>
          ) of times
        </Text>

        <div className={styles.button}>
          <Button
            onClick={this._handleClick}
            dataHook=\\"myNewComponent-button\\"
          >
            {buttonText}
          </Button>
        </div>
      </div>
    );
  }
}

export default MyNewComponent;
",
  "test-generated/src/MyNewComponent/MyNewComponent.meta.js": "import MyNewComponent from './MyNewComponent';
import Registry from '@ui-autotools/registry';

const metadata = Registry.getComponentMetadata(MyNewComponent);

metadata.exportedFrom({
  path: 'src/MyNewComponent/MyNewComponent.js',
});
",
  "test-generated/src/MyNewComponent/MyNewComponent.st.css": ":import {
  -st-from: \\"wix-ui-core/dist/src/themes/backoffice/colors.st.css\\";
  -st-named: G30,R30;
}

.root {
  text-align: center;
}

.number {
  -st-states: odd, even;
}

.number:even {
  background: value(G30);
}

.number:odd {
  background: value(R30);
}

.button {
  display: flex;
  justify-content: space-evenly;
}
",
  "test-generated/src/MyNewComponent/MyNewComponent.uni.driver.js": "import { baseUniDriverFactory } from 'wix-ui-test-utils/base-driver';

export const myNewComponentDriverFactory = base => {
  return {
    ...baseUniDriverFactory(base),

    /** Get the current count */
    getCountText: async () =>
      base.$('[data-hook=\\"myNewComponent-count\\"]').text(),

    /** Click the button */
    clickButton: async () =>
      base.$('[data-hook=\\"myNewComponent-button\\"]').click(),

    /** Get the button's text */
    getButtonText: async () =>
      base.$('[data-hook=\\"myNewComponent-button\\"]').text(),
  };
};
",
  "test-generated/src/MyNewComponent/docs/index.story.js": "import React from 'react';
import {
  header,
  tabs,
  tab,
  description,
  importExample,
  title,
  columns,
  divider,
  code as baseCode,
  playground,
  api,
  testkit,
} from 'wix-storybook-utils/Sections';

import { storySettings } from '../test/storySettings';
import allComponents from '../../../stories/utils/allComponents';

import MyNewComponent from '..';

const code = config => baseCode({ components: allComponents, ...config });

export default {
  category: storySettings.category,
  storyName: 'MyNewComponent',

  component: MyNewComponent,
  componentPath: '..',

  componentProps: {
    buttonText: 'Hello World!',
  },

  exampleProps: {
    // Put here presets of props, for more info:
    // https://github.com/wix/wix-ui/blob/master/packages/wix-storybook-utils/docs/usage.md#using-list
  },

  sections: [
    header({
      issueUrl: 'https://github.com/wix/wix-style-react/issues/new',
      sourceUrl:
        'https://github.com/wix/wix-style-react/tree/master/src/MyNewComponent/',
      component: <MyNewComponent buttonText=\\"Click me!\\" />,
    }),

    tabs([
      tab({
        title: 'Description',
        sections: [
          columns([
            description({
              title: 'Description',
              text:
                'This line here should briefly describe component in just a sentence or two. It should be short and easy to read.',
            }),
          ]),

          columns([
            importExample(
              \\"import MyNewComponent from 'wix-style-react/MyNewComponent';\\",
            ),
          ]),

          divider(),

          title('Examples'),

          columns([
            description({
              title: 'Simple Usage',
              text: 'A simple example with compact preview',
            }),

            code({
              compact: true,
              source: '<MyNewComponent buttonText=\\"Hello World!\\"/>',
            }),
          ]),

          code({
            title: 'Full Interactive Preview',
            description: 'A non compact version of same code example as above',
            source: '<MyNewComponent buttonText=\\"Hello World!\\"/>',
          }),
        ],
      }),

      ...[
        { title: 'API', sections: [api()] },
        { title: 'Testkit', sections: [testkit()] },
        { title: 'Playground', sections: [playground()] },
      ].map(tab),
    ]),
  ],
};
",
  "test-generated/src/MyNewComponent/index.js": "export { default } from './MyNewComponent.js';
",
  "test-generated/src/MyNewComponent/test/MyNewComponent.e2e.js": "import {
  waitForVisibilityOf,
  scrollToElement,
} from 'wix-ui-test-utils/protractor';

import { eyesItInstance } from '../../../test/utils/eyes-it';
import { createTestStoryUrl } from '../../../test/utils/storybook-helpers';
import { myNewComponentTestkitFactory } from '../../../testkit/protractor';
import { storySettings } from './storySettings';

const eyes = eyesItInstance();

describe('MyNewComponent', () => {
  const createStoryUrl = testName =>
    createTestStoryUrl({ ...storySettings, testName });

  const createDriver = async (dataHook = storySettings.dataHook) => {
    const driver = myNewComponentTestkitFactory({ dataHook });

    await waitForVisibilityOf(
      await driver.element(),
      \`Cannot find <MyNewComponent/> component with dataHook of \${dataHook}\`,
    );

    await scrollToElement(await driver.element());

    return driver;
  };

  const testStoryNames = storySettings.testStoryNames;

  eyes.it('should increase count when button clicked', async () => {
    await browser.get(createStoryUrl(testStoryNames.DEFAULT));
    const driver = await createDriver();
    await driver.clickButton();
    expect(await driver.getCountText()).toBe('You clicked this button odd number (1) of times');
  });
});
",
  "test-generated/src/MyNewComponent/test/MyNewComponent.private.uni.driver.js": "import { myNewComponentDriverFactory as publicDriverFactory } from '../MyNewComponent.uni.driver';

export const myNewComponentPrivateDriverFactory = base => {
  return {
    ...publicDriverFactory(base),

    // Add here driver methods that considered \\"private\\"
  };
};
",
  "test-generated/src/MyNewComponent/test/MyNewComponent.spec.js": "import React from 'react';
import { createRendererWithUniDriver, cleanup } from '../../../test/utils/unit';

import MyNewComponent from '../MyNewComponent';
import { myNewComponentPrivateDriverFactory } from './MyNewComponent.private.uni.driver';

describe('MyNewComponent', () => {
  const render = createRendererWithUniDriver(
    myNewComponentPrivateDriverFactory,
  );

  afterEach(() => {
    cleanup();
  });

  it('should render', async () => {
    const { driver } = render(<MyNewComponent />);

    expect(await driver.exists()).toBeTruthy();
    expect(await driver.getButtonText()).toEqual('Click me!');
  });

  it('should increment', async () => {
    const { driver } = render(<MyNewComponent />);

    await driver.clickButton();
    await driver.clickButton();

    expect(await driver.getCountText()).toEqual(
      'You clicked this button even number (2) of times',
    );
  });

  it('should allow changing the button text', async () => {
    const { driver } = render(<MyNewComponent buttonText=\\"Press me\\" />);

    expect(await driver.getButtonText()).toEqual('Press me');
  });
});
",
  "test-generated/src/MyNewComponent/test/MyNewComponent.visual.js": "import React from 'react';
import { storiesOf } from '@storybook/react';
import MyNewComponent from '../MyNewComponent';

const commonProps = {
  //use for repeated props across the tests (e.g. {buttonText: 'example'})
};

const tests = [
  {
    describe: 'sanity', // prop name (e.g. size)
    its: [
      {
        it: 'default', //prop vaiation (e.g. small)
        props: {
          // the simulation (e.g. {size: \\"small\\"})
        },
      },
    ],
  },
];

tests.forEach(({ describe, its }) => {
  its.forEach(({ it, props }) => {
    storiesOf(\`MyNewComponent/\${describe}\`, module).add(it, () => (
      <MyNewComponent {...commonProps} {...props} />
    ));
  });
});
",
  "test-generated/src/MyNewComponent/test/MyNewComponentStories.js": "import React from 'react';
import { storiesOf } from '@storybook/react';
import { getTestStoryKind } from '../../../stories/storiesHierarchy';
import { storySettings } from './storySettings';

import MyNewComponent from '..';

const TestStories = storiesOf(getTestStoryKind(storySettings), module);
const { testStoryNames, dataHook } = storySettings;

TestStories.add(testStoryNames.DEFAULT, () => (
  <MyNewComponent
    dataHook={dataHook}
    buttonText=\\"Press me for a surprise\\"
  />
));
",
  "test-generated/src/MyNewComponent/test/storySettings.js": "import { Category } from '../../../stories/storiesHierarchy';

export const storySettings = {
  category: Category.COMPONENTS,
  storyName: 'MyNewComponent',
  dataHook: 'story-my-new-component',
  testStoryNames: {
    DEFAULT: 'Default',
  },
};
",
  "test-generated/src/index.js": "// explicit test file
",
  "test-generated/stories/index.js": "",
  "test-generated/testkit/protractor.js": "// explicit test file
",
  "test-generated/testkit/puppeteer.js": "// explicit test file
",
  "test-generated/testkit/testkit-definitions.js": "// explicit test file
",
}
`;

exports[`copyTemplates should work as expected when description is provided 1`] = `
Object {
  "test-generated/src/MyNewComponent/MyNewComponent.js": "import React from 'react';
import PropTypes from 'prop-types';

import Text from '../Text';
import Button from '../Button';
import styles from './MyNewComponent.st.css';

/**
 * This is a very cool component, ya'll
 */
class MyNewComponent extends React.PureComponent {
  static displayName = 'MyNewComponent';

  static propTypes = {
    dataHook: PropTypes.string,

    /** Text for the button */
    buttonText: PropTypes.string,
  };

  static defaultProps = {
    buttonText: 'Click me!',
  };

  state = {
    count: 0,
  };

  _handleClick = () => {
    this.setState(({ count }) => ({
      count: count + 1,
    }));
  };

  render() {
    const { count } = this.state;
    const { dataHook, buttonText } = this.props;
    const isEven = count % 2 === 0;

    return (
      <div className={styles.root} data-hook={dataHook}>
        <Text dataHook=\\"myNewComponent-count\\">
          You clicked this button {isEven ? 'even' : 'odd'} number (
          <span
            {...styles('number', { even: isEven, odd: !isEven }, this.props)}
          >
            {count}
          </span>
          ) of times
        </Text>

        <div className={styles.button}>
          <Button
            onClick={this._handleClick}
            dataHook=\\"myNewComponent-button\\"
          >
            {buttonText}
          </Button>
        </div>
      </div>
    );
  }
}

export default MyNewComponent;
",
  "test-generated/src/MyNewComponent/MyNewComponent.meta.js": "import MyNewComponent from './MyNewComponent';
import Registry from '@ui-autotools/registry';

const metadata = Registry.getComponentMetadata(MyNewComponent);

metadata.exportedFrom({
  path: 'src/MyNewComponent/MyNewComponent.js',
});
",
  "test-generated/src/MyNewComponent/MyNewComponent.st.css": ":import {
  -st-from: \\"wix-ui-core/dist/src/themes/backoffice/colors.st.css\\";
  -st-named: G30,R30;
}

.root {
  text-align: center;
}

.number {
  -st-states: odd, even;
}

.number:even {
  background: value(G30);
}

.number:odd {
  background: value(R30);
}

.button {
  display: flex;
  justify-content: space-evenly;
}
",
  "test-generated/src/MyNewComponent/MyNewComponent.uni.driver.js": "import { baseUniDriverFactory } from 'wix-ui-test-utils/base-driver';

export const myNewComponentDriverFactory = base => {
  return {
    ...baseUniDriverFactory(base),

    /** Get the current count */
    getCountText: async () =>
      base.$('[data-hook=\\"myNewComponent-count\\"]').text(),

    /** Click the button */
    clickButton: async () =>
      base.$('[data-hook=\\"myNewComponent-button\\"]').click(),

    /** Get the button's text */
    getButtonText: async () =>
      base.$('[data-hook=\\"myNewComponent-button\\"]').text(),
  };
};
",
  "test-generated/src/MyNewComponent/docs/index.story.js": "import React from 'react';
import {
  header,
  tabs,
  tab,
  description,
  importExample,
  title,
  columns,
  divider,
  code as baseCode,
  playground,
  api,
  testkit,
} from 'wix-storybook-utils/Sections';

import { storySettings } from '../test/storySettings';
import allComponents from '../../../stories/utils/allComponents';

import MyNewComponent from '..';

const code = config => baseCode({ components: allComponents, ...config });

export default {
  category: storySettings.category,
  storyName: 'MyNewComponent',

  component: MyNewComponent,
  componentPath: '..',

  componentProps: {
    buttonText: 'Hello World!',
  },

  exampleProps: {
    // Put here presets of props, for more info:
    // https://github.com/wix/wix-ui/blob/master/packages/wix-storybook-utils/docs/usage.md#using-list
  },

  sections: [
    header({
      issueUrl: 'https://github.com/wix/wix-style-react/issues/new',
      sourceUrl:
        'https://github.com/wix/wix-style-react/tree/master/src/MyNewComponent/',
      component: <MyNewComponent buttonText=\\"Click me!\\" />,
    }),

    tabs([
      tab({
        title: 'Description',
        sections: [
          columns([
            description({
              title: 'Description',
              text:
                'This line here should briefly describe component in just a sentence or two. It should be short and easy to read.',
            }),
          ]),

          columns([
            importExample(
              \\"import MyNewComponent from 'wix-style-react/MyNewComponent';\\",
            ),
          ]),

          divider(),

          title('Examples'),

          columns([
            description({
              title: 'Simple Usage',
              text: 'A simple example with compact preview',
            }),

            code({
              compact: true,
              source: '<MyNewComponent buttonText=\\"Hello World!\\"/>',
            }),
          ]),

          code({
            title: 'Full Interactive Preview',
            description: 'A non compact version of same code example as above',
            source: '<MyNewComponent buttonText=\\"Hello World!\\"/>',
          }),
        ],
      }),

      ...[
        { title: 'API', sections: [api()] },
        { title: 'Testkit', sections: [testkit()] },
        { title: 'Playground', sections: [playground()] },
      ].map(tab),
    ]),
  ],
};
",
  "test-generated/src/MyNewComponent/index.js": "export { default } from './MyNewComponent.js';
",
  "test-generated/src/MyNewComponent/test/MyNewComponent.e2e.js": "import {
  waitForVisibilityOf,
  scrollToElement,
} from 'wix-ui-test-utils/protractor';

import { eyesItInstance } from '../../../test/utils/eyes-it';
import { createTestStoryUrl } from '../../../test/utils/storybook-helpers';
import { myNewComponentTestkitFactory } from '../../../testkit/protractor';
import { storySettings } from './storySettings';

const eyes = eyesItInstance();

describe('MyNewComponent', () => {
  const createStoryUrl = testName =>
    createTestStoryUrl({ ...storySettings, testName });

  const createDriver = async (dataHook = storySettings.dataHook) => {
    const driver = myNewComponentTestkitFactory({ dataHook });

    await waitForVisibilityOf(
      await driver.element(),
      \`Cannot find <MyNewComponent/> component with dataHook of \${dataHook}\`,
    );

    await scrollToElement(await driver.element());

    return driver;
  };

  const testStoryNames = storySettings.testStoryNames;

  eyes.it('should increase count when button clicked', async () => {
    await browser.get(createStoryUrl(testStoryNames.DEFAULT));
    const driver = await createDriver();
    await driver.clickButton();
    expect(await driver.getCountText()).toBe('You clicked this button odd number (1) of times');
  });
});
",
  "test-generated/src/MyNewComponent/test/MyNewComponent.private.uni.driver.js": "import { myNewComponentDriverFactory as publicDriverFactory } from '../MyNewComponent.uni.driver';

export const myNewComponentPrivateDriverFactory = base => {
  return {
    ...publicDriverFactory(base),

    // Add here driver methods that considered \\"private\\"
  };
};
",
  "test-generated/src/MyNewComponent/test/MyNewComponent.spec.js": "import React from 'react';
import { createRendererWithUniDriver, cleanup } from '../../../test/utils/unit';

import MyNewComponent from '../MyNewComponent';
import { myNewComponentPrivateDriverFactory } from './MyNewComponent.private.uni.driver';

describe('MyNewComponent', () => {
  const render = createRendererWithUniDriver(
    myNewComponentPrivateDriverFactory,
  );

  afterEach(() => {
    cleanup();
  });

  it('should render', async () => {
    const { driver } = render(<MyNewComponent />);

    expect(await driver.exists()).toBeTruthy();
    expect(await driver.getButtonText()).toEqual('Click me!');
  });

  it('should increment', async () => {
    const { driver } = render(<MyNewComponent />);

    await driver.clickButton();
    await driver.clickButton();

    expect(await driver.getCountText()).toEqual(
      'You clicked this button even number (2) of times',
    );
  });

  it('should allow changing the button text', async () => {
    const { driver } = render(<MyNewComponent buttonText=\\"Press me\\" />);

    expect(await driver.getButtonText()).toEqual('Press me');
  });
});
",
  "test-generated/src/MyNewComponent/test/MyNewComponent.visual.js": "import React from 'react';
import { storiesOf } from '@storybook/react';
import MyNewComponent from '../MyNewComponent';

const commonProps = {
  //use for repeated props across the tests (e.g. {buttonText: 'example'})
};

const tests = [
  {
    describe: 'sanity', // prop name (e.g. size)
    its: [
      {
        it: 'default', //prop vaiation (e.g. small)
        props: {
          // the simulation (e.g. {size: \\"small\\"})
        },
      },
    ],
  },
];

tests.forEach(({ describe, its }) => {
  its.forEach(({ it, props }) => {
    storiesOf(\`MyNewComponent/\${describe}\`, module).add(it, () => (
      <MyNewComponent {...commonProps} {...props} />
    ));
  });
});
",
  "test-generated/src/MyNewComponent/test/MyNewComponentStories.js": "import React from 'react';
import { storiesOf } from '@storybook/react';
import { getTestStoryKind } from '../../../stories/storiesHierarchy';
import { storySettings } from './storySettings';

import MyNewComponent from '..';

const TestStories = storiesOf(getTestStoryKind(storySettings), module);
const { testStoryNames, dataHook } = storySettings;

TestStories.add(testStoryNames.DEFAULT, () => (
  <MyNewComponent
    dataHook={dataHook}
    buttonText=\\"Press me for a surprise\\"
  />
));
",
  "test-generated/src/MyNewComponent/test/storySettings.js": "import { Category } from '../../../stories/storiesHierarchy';

export const storySettings = {
  category: Category.COMPONENTS,
  storyName: 'MyNewComponent',
  dataHook: 'story-my-new-component',
  testStoryNames: {
    DEFAULT: 'Default',
  },
};
",
  "test-generated/src/index.js": "// explicit test file
",
  "test-generated/stories/index.js": "",
  "test-generated/testkit/protractor.js": "// explicit test file
",
  "test-generated/testkit/puppeteer.js": "// explicit test file
",
  "test-generated/testkit/testkit-definitions.js": "// explicit test file
",
}
`;
